File: smart-account-demo/shared/src/commonMain/kotlin/com/soneso/smartdemo/ui/screens/MainScreen.kt
package com.soneso.smartdemo.ui.screens

/**
 * Main screen: wallet status dashboard, navigation, and activity log.
 * SDK initialization and balance refresh are handled by MainScreenFlow.
 *
 * When a wallet is connected but not yet deployed on-chain, a warning card is shown
 * with a "Deploy Now" button. Navigation buttons for features that require an on-chain
 * contract (Transfer, Context Rules, Account Signers) are disabled until deployment.
 */

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Notifications
import androidx.compose.material3.Badge
import androidx.compose.material3.BadgedBox
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Scaffold
import androidx.compose.material3.SnackbarHost
import androidx.compose.material3.SnackbarHostState
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import cafe.adriel.voyager.core.screen.Screen
import cafe.adriel.voyager.navigator.LocalNavigator
import cafe.adriel.voyager.navigator.currentOrThrow
import com.soneso.smartdemo.flows.deployPendingAndProvision
import com.soneso.smartdemo.flows.disconnect
import com.soneso.smartdemo.flows.initializeKit
import com.soneso.smartdemo.flows.refreshBalances
import com.soneso.smartdemo.platform.Clipboard
import com.soneso.smartdemo.platform.getClipboard
import com.soneso.smartdemo.state.ActivityLogState
import com.soneso.smartdemo.state.DemoState
import com.soneso.smartdemo.state.LogEntry
import com.soneso.smartdemo.state.LogLevel
import kotlin.time.ExperimentalTime
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import kotlinx.datetime.TimeZone
import kotlinx.datetime.toLocalDateTime

class MainScreen : Screen {

    @OptIn(ExperimentalMaterial3Api::class)
    @Composable
    override fun Content() {
        val navigator = LocalNavigator.currentOrThrow
        val scope = rememberCoroutineScope()
        val snackbarHostState = remember { SnackbarHostState() }
        val clipboard = remember { getClipboard() }

        // Controls the Refresh button loading state
        var isRefreshingBalance by remember { mutableStateOf(false) }

        // Controls the Deploy Now button in the undeployed warning card
        var isDeploying by remember { mutableStateOf(false) }
        var deployErrorMessage by remember { mutableStateOf<String?>(null) }

        // Auto-initialize the SDK when platform providers become available
        LaunchedEffect(DemoState.webauthnProvider, DemoState.storage) {
            if (DemoState.kit == null && DemoState.webauthnProvider != null) {
                try {
                    initializeKit(DemoState.webauthnProvider, DemoState.storage)
                } catch (e: Exception) {
                    ActivityLogState.error("Failed to initialize SDK: ${e.message}")
                }
            }
        }

        Scaffold(
            snackbarHost = { SnackbarHost(snackbarHostState) },
            topBar = {
                TopAppBar(
                    title = {
                        Column {
                            Text("Stellar Smart Account Demo")
                            Text(
                                text = "Testnet",
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.onPrimary
                            )
                        }
                    },
                    actions = {
                        InboxBell(
                            pendingCount = DemoState.pendingRequestCount,
                            onClick = { navigator.push(ApprovalInboxScreen()) }
                        )
                    },
                    colors = TopAppBarDefaults.topAppBarColors(
                        containerColor = MaterialTheme.colorScheme.primary,
                        titleContentColor = MaterialTheme.colorScheme.onPrimary,
                        actionIconContentColor = MaterialTheme.colorScheme.onPrimary
                    )
                )
            }
        ) { padding ->
            Column(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(padding)
                    .padding(16.dp)
                    .verticalScroll(rememberScrollState()),
                verticalArrangement = Arrangement.spacedBy(16.dp)
            ) {
                // 1. Wallet Status Section
                Card(
                    modifier = Modifier.fillMaxWidth(),
                    colors = CardDefaults.cardColors(
                        containerColor = MaterialTheme.colorScheme.primaryContainer
                    )
                ) {
                    Column(modifier = Modifier.padding(16.dp)) {
                        Text(
                            text = "Wallet Status",
                            style = MaterialTheme.typography.titleMedium,
                            fontWeight = FontWeight.Bold,
                            color = MaterialTheme.colorScheme.onPrimaryContainer
                        )

                        Spacer(modifier = Modifier.height(12.dp))

                        if (!DemoState.isConnected) {
                            Text(
                                text = "No wallet connected",
                                style = MaterialTheme.typography.bodyMedium,
                                color = MaterialTheme.colorScheme.onPrimaryContainer
                            )

                            Spacer(modifier = Modifier.height(12.dp))

                            Row(
                                modifier = Modifier.fillMaxWidth(),
                                horizontalArrangement = Arrangement.spacedBy(8.dp)
                            ) {
                                Button(
                                    onClick = { navigator.push(WalletCreationScreen()) },
                                    modifier = Modifier.weight(1f)
                                ) {
                                    Text("Create Wallet")
                                }
                                Button(
                                    onClick = { navigator.push(WalletConnectionScreen()) },
                                    modifier = Modifier.weight(1f)
                                ) {
                                    Text("Connect Wallet")
                                }
                            }
                        } else {
                            // Contract Address
                            Text(
                                text = "Contract Address:",
                                style = MaterialTheme.typography.labelMedium,
                                color = MaterialTheme.colorScheme.onPrimaryContainer
                            )
                            Row(
                                verticalAlignment = Alignment.CenterVertically
                            ) {
                                Text(
                                    text = DemoState.contractId?.let { "${it.take(12)}...${it.takeLast(12)}" } ?: "Unknown",
                                    style = MaterialTheme.typography.bodyMedium,
                                    fontFamily = FontFamily.Monospace,
                                    color = MaterialTheme.colorScheme.onPrimaryContainer,
                                    modifier = Modifier.weight(1f)
                                )
                                if (DemoState.contractId != null) {
                                    OutlinedButton(
                                        onClick = {
                                            scope.launch {
                                                clipboard.copyToClipboard(DemoState.contractId!!)
                                                snackbarHostState.showSnackbar("Contract address copied")
                                            }
                                        },
                                        modifier = Modifier.padding(start = 8.dp)
                                    ) {
                                        Text("Copy", style = MaterialTheme.typography.labelSmall)
                                    }
                                }
                            }

                            Spacer(modifier = Modifier.height(8.dp))

                            // Credential ID
                            Text(
                                text = "Credential ID:",
                                style = MaterialTheme.typography.labelMedium,
                                color = MaterialTheme.colorScheme.onPrimaryContainer
                            )
                            Text(
                                text = DemoState.credentialId ?: "Unknown",
                                style = MaterialTheme.typography.bodyMedium,
                                fontFamily = FontFamily.Monospace,
                                color = MaterialTheme.colorScheme.onPrimaryContainer
                            )

                            Spacer(modifier = Modifier.height(8.dp))

                            // Undeployed warning card
                            if (!DemoState.isDeployed) {
                                UndeployedWarningCard(
                                    isDeploying = isDeploying,
                                    deployErrorMessage = deployErrorMessage,
                                    onDeploy = {
                                        val credId = DemoState.credentialId ?: return@UndeployedWarningCard
                                        scope.launch {
                                            isDeploying = true
                                            deployErrorMessage = null
                                            try {
                                                val result = deployPendingAndProvision(credId)
                                                if (!result.success) {
                                                    deployErrorMessage = "Deployment failed: ${result.error}"
                                                    ActivityLogState.error("Deployment failed: ${result.error}")
                                                }
                                            } catch (e: Throwable) {
                                                deployErrorMessage = "Deployment failed: ${e.message}"
                                                ActivityLogState.error("Deployment failed: ${e.message}")
                                            } finally {
                                                isDeploying = false
                                            }
                                        }
                                    }
                                )

                                Spacer(modifier = Modifier.height(8.dp))
                            }

                            // Balances (only show when deployed)
                            if (DemoState.isDeployed) {
                                Text(
                                    text = "Balance:",
                                    style = MaterialTheme.typography.labelMedium,
                                    color = MaterialTheme.colorScheme.onPrimaryContainer
                                )
                                Row(
                                    verticalAlignment = Alignment.CenterVertically,
                                    horizontalArrangement = Arrangement.spacedBy(8.dp)
                                ) {
                                    Column(modifier = Modifier.weight(1f)) {
                                        Text(
                                            text = "${DemoState.balance ?: "Loading..."} XLM",
                                            style = MaterialTheme.typography.bodyMedium,
                                            fontWeight = FontWeight.Bold,
                                            color = MaterialTheme.colorScheme.onPrimaryContainer
                                        )
                                        Text(
                                            text = "${DemoState.demoTokenBalance ?: "0.0"} DEMO",
                                            style = MaterialTheme.typography.bodyMedium,
                                            fontWeight = FontWeight.Bold,
                                            color = MaterialTheme.colorScheme.onPrimaryContainer
                                        )
                                    }
                                    OutlinedButton(
                                        onClick = {
                                            scope.launch {
                                                isRefreshingBalance = true
                                                try {
                                                    refreshBalances()
                                                } catch (e: Throwable) {
                                                    ActivityLogState.error("Failed to refresh balance: ${e.message}")
                                                } finally {
                                                    isRefreshingBalance = false
                                                }
                                            }
                                        },
                                        enabled = !isRefreshingBalance
                                    ) {
                                        Text(if (isRefreshingBalance) "Refreshing..." else "Refresh")
                                    }
                                }
                            }

                            Spacer(modifier = Modifier.height(12.dp))

                            // Navigation buttons
                            Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                                Row(
                                    modifier = Modifier.fillMaxWidth(),
                                    horizontalArrangement = Arrangement.spacedBy(8.dp)
                                ) {
                                    Button(
                                        onClick = { navigator.push(ContextRulesScreen()) },
                                        modifier = Modifier.weight(1f),
                                        enabled = DemoState.isDeployed,
                                        colors = if (DemoState.isDeployed) {
                                            ButtonDefaults.buttonColors()
                                        } else {
                                            ButtonDefaults.buttonColors(
                                                disabledContainerColor = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.12f),
                                                disabledContentColor = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.38f)
                                            )
                                        }
                                    ) {
                                        Text("Context Rules")
                                    }
                                    Button(
                                        onClick = { navigator.push(TransferScreen()) },
                                        modifier = Modifier.weight(1f),
                                        enabled = DemoState.isDeployed,
                                        colors = if (DemoState.isDeployed) {
                                            ButtonDefaults.buttonColors()
                                        } else {
                                            ButtonDefaults.buttonColors(
                                                disabledContainerColor = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.12f),
                                                disabledContentColor = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.38f)
                                            )
                                        }
                                    ) {
                                        Text("Transfer")
                                    }
                                }
                                Row(
                                    modifier = Modifier.fillMaxWidth(),
                                    horizontalArrangement = Arrangement.spacedBy(8.dp)
                                ) {
                                    Button(
                                        onClick = { navigator.push(ApproveScreen()) },
                                        modifier = Modifier.weight(1f),
                                        enabled = DemoState.isDeployed,
                                        colors = if (DemoState.isDeployed) {
                                            ButtonDefaults.buttonColors()
                                        } else {
                                            ButtonDefaults.buttonColors(
                                                disabledContainerColor = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.12f),
                                                disabledContentColor = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.38f)
                                            )
                                        }
                                    ) {
                                        Text("Approve")
                                    }
                                    Button(
                                        onClick = { navigator.push(KnownSignersScreen()) },
                                        modifier = Modifier.weight(1f),
                                        enabled = DemoState.isDeployed,
                                        colors = if (DemoState.isDeployed) {
                                            ButtonDefaults.buttonColors()
                                        } else {
                                            ButtonDefaults.buttonColors(
                                                disabledContainerColor = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.12f),
                                                disabledContentColor = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.38f)
                                            )
                                        }
                                    ) {
                                        Text("Account Signers")
                                    }
                                }
                                OutlinedButton(
                                    onClick = {
                                        scope.launch {
                                            try {
                                                disconnect()
                                            } catch (e: Throwable) {
                                                ActivityLogState.error("Disconnect failed: ${e.message}")
                                            }
                                        }
                                    },
                                    modifier = Modifier.fillMaxWidth(),
                                    enabled = !isDeploying
                                ) {
                                    Text("Disconnect")
                                }
                            }
                        }
                    }
                }

                // 2. Activity Log Section
                Card(
                    modifier = Modifier.fillMaxWidth(),
                    colors = CardDefaults.cardColors(
                        containerColor = MaterialTheme.colorScheme.surface
                    )
                ) {
                    Column(modifier = Modifier.padding(16.dp)) {
                        Row(
                            modifier = Modifier.fillMaxWidth(),
                            horizontalArrangement = Arrangement.SpaceBetween,
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            Text(
                                text = "Activity Log (${ActivityLogState.entries.size})",
                                style = MaterialTheme.typography.titleMedium,
                                fontWeight = FontWeight.Bold
                            )
                            OutlinedButton(
                                onClick = { ActivityLogState.clear() },
                                enabled = ActivityLogState.entries.isNotEmpty()
                            ) {
                                Text("Clear")
                            }
                        }

                        Spacer(modifier = Modifier.height(12.dp))

                        if (ActivityLogState.entries.isEmpty()) {
                            Text(
                                text = "No activity yet",
                                style = MaterialTheme.typography.bodyMedium,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        } else {
                            Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                                ActivityLogState.entries.forEach { entry ->
                                    LogEntryRow(
                                        entry = entry,
                                        clipboard = clipboard,
                                        snackbarHostState = snackbarHostState,
                                        scope = scope
                                    )
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    /**
     * Badged inbox bell shown in the top app bar. The badge displays the number of pending
     * agent escalations for the connected account; tapping opens the approval inbox.
     */
    @Composable
    private fun InboxBell(pendingCount: Int, onClick: () -> Unit) {
        IconButton(onClick = onClick) {
            BadgedBox(
                badge = {
                    if (pendingCount > 0) {
                        Badge {
                            Text(if (pendingCount > 99) "99+" else pendingCount.toString())
                        }
                    }
                }
            ) {
                Icon(
                    imageVector = Icons.Default.Notifications,
                    contentDescription = "Approval inbox"
                )
            }
        }
    }

    @Composable
    private fun UndeployedWarningCard(
        isDeploying: Boolean,
        deployErrorMessage: String?,
        onDeploy: () -> Unit
    ) {
        val warningContainerColor = Color(0xFFFFF3E0) // Orange 50
        val warningContentColor = Color(0xFFE65100) // Orange 900

        Card(
            modifier = Modifier.fillMaxWidth(),
            colors = CardDefaults.cardColors(
                containerColor = warningContainerColor
            )
        ) {
            Column(
                modifier = Modifier.padding(12.dp),
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                Text(
                    text = "Wallet Not Deployed",
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.Bold,
                    color = warningContentColor
                )
                Text(
                    text = "Your passkey is registered but the smart account contract has not been deployed to the network. Deploy it to start using your wallet.",
                    style = MaterialTheme.typography.bodySmall,
                    color = warningContentColor
                )

                if (deployErrorMessage != null) {
                    Text(
                        text = deployErrorMessage,
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.error
                    )
                }

                if (isDeploying) {
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.Center,
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        CircularProgressIndicator(
                            modifier = Modifier.size(20.dp),
                            strokeWidth = 2.dp,
                            color = warningContentColor
                        )
                        Spacer(modifier = Modifier.width(8.dp))
                        Text(
                            text = "Deploying contract...",
                            style = MaterialTheme.typography.bodySmall,
                            color = warningContentColor
                        )
                    }
                }

                Button(
                    onClick = onDeploy,
                    modifier = Modifier.fillMaxWidth(),
                    enabled = !isDeploying,
                    colors = ButtonDefaults.buttonColors(
                        containerColor = warningContentColor,
                        contentColor = Color.White
                    )
                ) {
                    Text(if (isDeploying) "Deploying..." else "Deploy Now")
                }
            }
        }
    }

    @OptIn(ExperimentalTime::class)
    @Composable
    private fun LogEntryRow(
        entry: LogEntry,
        clipboard: Clipboard,
        snackbarHostState: SnackbarHostState,
        scope: CoroutineScope
    ) {
        val timeString = remember(entry.timestamp) {
            val localTime = entry.timestamp.toLocalDateTime(TimeZone.currentSystemDefault())
            val h = localTime.hour.toString().padStart(2, '0')
            val m = localTime.minute.toString().padStart(2, '0')
            val s = localTime.second.toString().padStart(2, '0')
            "$h:$m:$s"
        }

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .clickable {
                    scope.launch {
                        clipboard.copyToClipboard(entry.message)
                        snackbarHostState.showSnackbar("Log message copied to clipboard")
                    }
                }
                .padding(vertical = 2.dp),
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text(
                text = timeString,
                style = MaterialTheme.typography.bodySmall,
                fontFamily = FontFamily.Monospace,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.width(70.dp)
            )

            Surface(
                color = when (entry.level) {
                    LogLevel.INFO -> Color(0xFF2196F3)
                    LogLevel.SUCCESS -> Color(0xFF4CAF50)
                    LogLevel.ERROR -> Color(0xFFF44336)
                },
                shape = MaterialTheme.shapes.small
            ) {
                Text(
                    text = when (entry.level) {
                        LogLevel.INFO -> "INFO"
                        LogLevel.SUCCESS -> "OK"
                        LogLevel.ERROR -> "ERR"
                    },
                    modifier = Modifier.padding(horizontal = 6.dp, vertical = 2.dp),
                    style = MaterialTheme.typography.labelSmall,
                    color = Color.White
                )
            }

            Text(
                text = entry.message,
                style = MaterialTheme.typography.bodySmall,
                modifier = Modifier.weight(1f)
            )
        }
    }
}

File: smart-account-demo/macosApp/StellarSmartDemo/Views/MainScreen.swift
//
//  MainScreen.swift
//  StellarSmartDemo
//
//  Copyright (c) 2026 Soneso. All rights reserved.
//

import SwiftUI
import shared

// MARK: - MainScreen

/// Root screen of the Smart Account Demo.
///
/// Mirrors the Compose `MainScreen` exactly. Shows either a disconnected wallet state
/// (with Create/Connect actions) or a connected wallet state (with contract address,
/// credential ID, balances, and navigation buttons). An activity log section is always
/// visible below the wallet status card.
///
/// When a wallet is connected but not yet deployed on-chain, a warning card is shown
/// with a "Deploy Now" button. Navigation buttons for features that require an on-chain
/// contract (Transfer, Approve, Context Rules, Account Signers) are disabled until deployment.
///
/// SDK initialization runs once via `.task` on first appear. After every bridge operation
/// `appState.sync(from:)` is called on the `MainActor` to propagate Kotlin state changes
/// into the observable SwiftUI state.
struct MainScreen: View {

    // MARK: - Environment

    @EnvironmentObject var bridgeWrapper: MacOSBridgeWrapper
    @EnvironmentObject var appState: AppState

    // MARK: - Properties

    @ObservedObject var toastManager: ToastManager

    // MARK: - Local state

    @State private var isRefreshingBalance = false
    @State private var hasInitialized = false
    @State private var isInitializingKit = false
    @State private var initErrorMessage: String? = nil
    @State private var isDeploying = false
    @State private var deployErrorMessage: String? = nil

    // MARK: - Init

    init(toastManager: ToastManager) {
        self.toastManager = toastManager
    }

    // MARK: - Body

    var body: some View {
        ScrollView {
            VStack(spacing: 16) {
                initErrorSection
                walletStatusSection
                activityLogSection
            }
            .padding(16)
        }
        .background(Material3Colors.background)
        .frame(minWidth: 550, minHeight: 700)
        .navigationToolbar(
            title: "Stellar Smart Account Demo",
            subtitle: "Testnet",
            showBackButton: false
        )
        .task {
            guard !hasInitialized else { return }
            hasInitialized = true
            await initializeKitOnce()
        }
    }

    // MARK: - Init Error Section

    /// Shown when SDK kit initialization failed. All wallet actions depend on the kit,
    /// so the card offers a Retry that re-runs initialization.
    @ViewBuilder
    private var initErrorSection: some View {
        if let message = initErrorMessage {
            VStack(alignment: .leading, spacing: 8) {
                Text("Initialization Failed")
                    .font(.system(size: 14, weight: .bold))
                    .foregroundColor(Material3Colors.onErrorContainer)

                Text(message)
                    .font(.system(size: 13))
                    .foregroundColor(Material3Colors.onErrorContainer)
                    .textSelection(.enabled)

                if isInitializingKit {
                    HStack(spacing: 6) {
                        ProgressView()
                            .controlSize(.small)
                            .tint(Material3Colors.error)
                        Text("Initializing...")
                            .font(.system(size: 13))
                            .foregroundColor(Material3Colors.onErrorContainer)
                    }
                }

                Button(action: {
                    Task { await initializeKitOnce() }
                }) {
                    Text(isInitializingKit ? "Retrying..." : "Retry")
                        .font(.system(size: 14, weight: .medium))
                        .frame(maxWidth: .infinity)
                        .frame(height: 36)
                        .foregroundColor(.white)
                        .background(isInitializingKit
                            ? Material3Colors.error.opacity(0.5)
                            : Material3Colors.error)
                        .cornerRadius(6)
                }
                .buttonStyle(.plain)
                .disabled(isInitializingKit)
            }
            .padding(12)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(Material3Colors.errorContainer)
            .cornerRadius(8)
        }
    }

    // MARK: - Wallet Status Section

    @ViewBuilder
    private var walletStatusSection: some View {
        InfoCard(title: "Wallet Status", color: .primary) {
            if appState.isConnected {
                connectedWalletContent
            } else {
                disconnectedWalletContent
            }
        }
    }

    // MARK: - Disconnected State

    private var disconnectedWalletContent: some View {
        VStack(alignment: .leading, spacing: 12) {
            // Info description
            Text(
                "A smart account is a Soroban contract wallet controlled by passkey credentials. " +
                "Create a new wallet to register a passkey and deploy a contract, or connect an existing one."
            )
            .font(.callout)
            .foregroundColor(Material3Colors.onPrimaryContainer)

            // Status badge
            HStack(spacing: 8) {
                Image(systemName: "info.circle.fill")
                    .font(.system(size: 14))
                    .foregroundColor(Material3Colors.primary)
                Text("No wallet connected")
                    .font(.callout)
                    .foregroundColor(Material3Colors.onPrimaryContainer)
            }
            .padding(10)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(Material3Colors.primaryContainer.opacity(0.6))
            .overlay(
                RoundedRectangle(cornerRadius: 6)
                    .stroke(Material3Colors.primary.opacity(0.3), lineWidth: 1)
            )
            .cornerRadius(6)

            // Action buttons
            HStack(spacing: 8) {
                NavigationLink(destination: WalletCreationScreen(toastManager: toastManager)) {
                    PrimaryButtonLabel(
                        text: "Create Wallet",
                        style: .filled(foreground: .white, background: Material3Colors.primary)
                    )
                }
                .buttonStyle(.plain)

                NavigationLink(destination: WalletConnectionScreen(toastManager: toastManager)) {
                    PrimaryButtonLabel(
                        text: "Connect Wallet",
                        style: .filled(foreground: .white, background: Material3Colors.primary)
                    )
                }
                .buttonStyle(.plain)
            }
        }
    }

    // MARK: - Connected State

    private var connectedWalletContent: some View {
        VStack(alignment: .leading, spacing: 8) {

            // Contract Address
            if let contractId = appState.contractId {
                CopyableField(
                    label: "Contract Address",
                    value: KotlinInterop.truncateAddress(contractId, prefixLength: 12, suffixLength: 12),
                    textColor: Material3Colors.onPrimaryContainer,
                    labelColor: Material3Colors.onPrimaryContainer.opacity(0.7),
                    monospace: true,
                    onCopy: {
                        ClipboardHelper.copy(contractId)
                        toastManager.show("Contract address copied")
                    }
                )
            } else {
                unknownValueRow(label: "Contract Address")
            }

            Divider()
                .background(Material3Colors.onPrimaryContainer.opacity(0.15))

            // Credential ID
            if let credentialId = appState.credentialId {
                CopyableField(
                    label: "Credential ID",
                    value: credentialId,
                    textColor: Material3Colors.onPrimaryContainer,
                    labelColor: Material3Colors.onPrimaryContainer.opacity(0.7),
                    monospace: true,
                    onCopy: {
                        ClipboardHelper.copy(credentialId)
                        toastManager.show("Credential ID copied")
                    }
                )
            } else {
                unknownValueRow(label: "Credential ID")
            }

            Divider()
                .background(Material3Colors.onPrimaryContainer.opacity(0.15))

            // Undeployed warning card
            if !appState.isDeployed {
                undeployedWarningCard

                Divider()
                    .background(Material3Colors.onPrimaryContainer.opacity(0.15))
            }

            // Balances + Refresh (only when deployed)
            if appState.isDeployed {
                HStack(alignment: .top) {
                    BalanceRows(
                        label: "Balance",
                        labelColor: Material3Colors.onPrimaryContainer.opacity(0.7),
                        values: [
                            "\(appState.xlmBalance ?? "Loading...") XLM",
                            "\(appState.demoTokenBalance ?? "0.0") DEMO"
                        ],
                        valueFont: .callout.bold(),
                        valueColor: Material3Colors.onPrimaryContainer
                    )

                    Spacer()

                    Button(action: refreshBalance) {
                        HStack(spacing: 6) {
                            if isRefreshingBalance {
                                ProgressView()
                                    .controlSize(.small)
                                    .tint(Material3Colors.primary)
                                Text("Refreshing...")
                                    .font(.caption)
                                    .foregroundColor(Material3Colors.primary)
                            } else {
                                Text("Refresh")
                                    .font(.caption)
                                    .foregroundColor(Material3Colors.primary)
                            }
                        }
                        .padding(.horizontal, 12)
                        .padding(.vertical, 6)
                        .overlay(
                            RoundedRectangle(cornerRadius: 6)
                                .stroke(Material3Colors.primary, lineWidth: 1)
                        )
                    }
                    .buttonStyle(.plain)
                    .disabled(isRefreshingBalance)
                    .opacity(isRefreshingBalance ? 0.5 : 1.0)
                }

                Divider()
                    .background(Material3Colors.onPrimaryContainer.opacity(0.15))
                    .padding(.top, 4)
            }

            // Navigation buttons
            VStack(spacing: 8) {
                HStack(spacing: 8) {
                    navButton("Context Rules", destination: ContextRulesScreen(toastManager: toastManager))
                    navButton("Transfer", destination: TransferScreen(toastManager: toastManager))
                }

                HStack(spacing: 8) {
                    navButton("Approve", destination: ApproveScreen(toastManager: toastManager))
                    navButton("Account Signers", destination: KnownSignersScreen())
                }

                // Disconnect -- outlined with error-ish styling (disabled during deployment)
                LoadingButton(
                    action: disconnectWallet,
                    isLoading: false,
                    isEnabled: !isDeploying,
                    text: "Disconnect",
                    loadingText: "Disconnecting...",
                    style: .outlined
                )
            }
        }
    }

    /// Filled navigation button for the connected-wallet feature screens.
    /// Enabled only while the wallet contract is deployed on-chain.
    private func navButton<Destination: View>(_ title: String, destination: Destination) -> some View {
        NavigationLink(destination: destination) {
            PrimaryButtonLabel(
                text: title,
                style: .filled(
                    foreground: appState.isDeployed ? .white : .gray,
                    background: appState.isDeployed ? Material3Colors.primary : Color.gray.opacity(0.3)
                )
            )
        }
        .buttonStyle(.plain)
        .disabled(!appState.isDeployed)
    }

    /// Labeled row shown when a connected-wallet value is not available.
    private func unknownValueRow(label: String) -> some View {
        VStack(alignment: .leading, spacing: 2) {
            Text(label)
                .font(.caption)
                .fontWeight(.bold)
                .foregroundColor(Material3Colors.onPrimaryContainer.opacity(0.7))
            Text("Unknown")
                .font(.system(.callout, design: .monospaced))
                .foregroundColor(Material3Colors.onPrimaryContainer)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    // MARK: - Undeployed Warning Card

    private var undeployedWarningCard: some View {
        let warningColor = Material3Colors.warningText
        let warningBg = Material3Colors.warningBackground

        return VStack(alignment: .leading, spacing: 8) {
            Text("Wallet Not Deployed")
                .font(.system(size: 14, weight: .bold))
                .foregroundColor(warningColor)

            Text(
                "Your passkey is registered but the smart account contract has not been " +
                "deployed to the network. Deploy it to start using your wallet."
            )
            .font(.system(size: 13))
            .foregroundColor(warningColor)

            if let error = deployErrorMessage {
                Text(error)
                    .font(.system(size: 12))
                    .foregroundColor(Material3Colors.error)
            }

            if isDeploying {
                HStack(spacing: 6) {
                    ProgressView()
                        .controlSize(.small)
                        .tint(warningColor)
                    Text("Deploying contract...")
                        .font(.system(size: 13))
                        .foregroundColor(warningColor)
                }
            }

            Button(action: deployFromMainScreen) {
                Text(isDeploying ? "Deploying..." : "Deploy Now")
                    .font(.system(size: 14, weight: .medium))
                    .frame(maxWidth: .infinity)
                    .frame(height: 36)
                    .foregroundColor(.white)
                    .background(isDeploying ? warningColor.opacity(0.5) : warningColor)
                    .cornerRadius(6)
            }
            .buttonStyle(.plain)
            .disabled(isDeploying)
        }
        .padding(12)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(warningBg)
        .cornerRadius(8)
    }

    // MARK: - Activity Log Section

    private var activityLogSection: some View {
        VStack(alignment: .leading, spacing: 0) {
            ActivityLogView(
                entries: appState.activityLog,
                onClear: {
                    appState.clearActivityLog(bridge: bridgeWrapper.bridge)
                },
                onCopyEntry: { _ in
                    toastManager.show("Log message copied to clipboard")
                }
            )
        }
        .padding(16)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Material3Colors.surface)
        .cornerRadius(8)
        .shadow(color: Material3Colors.cardShadow, radius: 2, x: 0, y: 1)
    }

    // MARK: - Actions

    private func initializeKitOnce() async {
        // Both call sites (the one-shot .task and the Retry button) run on the main
        // actor, so this synchronous guard prevents concurrent initialization runs.
        guard !isInitializingKit else { return }
        isInitializingKit = true
        let bridge = bridgeWrapper.bridge
        do {
            try await bridge.initializeKit()
            await MainActor.run {
                bridgeWrapper.isKitInitialized = true
                initErrorMessage = nil
            }
        } catch {
            // The bridge logs the failure to the activity log; the card adds a Retry path.
            await MainActor.run {
                initErrorMessage = "Failed to initialize: \(error.localizedDescription)"
            }
        }
        await MainActor.run {
            appState.sync(from: bridge)
            isInitializingKit = false
        }
    }

    private func refreshBalance() {
        let bridge = bridgeWrapper.bridge
        isRefreshingBalance = true
        Task {
            do {
                try await bridge.refreshBalances()
            } catch {
                await MainActor.run {
                    ActivityLogState.shared.error(
                        message: "Failed to refresh balance: \(error.localizedDescription)"
                    )
                    appState.syncActivityLog(from: bridge)
                }
            }
            await MainActor.run {
                appState.sync(from: bridge)
                isRefreshingBalance = false
            }
        }
    }

    private func disconnectWallet() {
        let bridge = bridgeWrapper.bridge
        Task {
            do {
                try await bridge.disconnect()
            } catch {
                await MainActor.run {
                    ActivityLogState.shared.error(
                        message: "Disconnect failed: \(error.localizedDescription)"
                    )
                    appState.syncActivityLog(from: bridge)
                }
            }
            await MainActor.run {
                appState.sync(from: bridge)
            }
        }
    }

    private func deployFromMainScreen() {
        guard let credentialId = appState.credentialId else { return }
        let bridge = bridgeWrapper.bridge
        isDeploying = true
        deployErrorMessage = nil

        Task {
            do {
                let result = try await bridge.deployPendingAndProvision(
                    credentialId: credentialId,
                    onProgress: { _ in }
                )
                await MainActor.run {
                    appState.sync(from: bridge)
                    if !result.success {
                        deployErrorMessage = "Deployment failed: \(result.error ?? "Unknown error")"
                    }
                    isDeploying = false
                    if result.success {
                        toastManager.show("Wallet deployed successfully")
                    }
                }
            } catch {
                await MainActor.run {
                    deployErrorMessage = "Deployment failed: \(error.localizedDescription)"
                    isDeploying = false
                    appState.syncActivityLog(from: bridge)
                }
            }
        }
    }
}
